src/lib.rs

```diff
@@ -190,19 +182,16 @@ impl<const N: usize, A: Alphabet<N>> TryFrom<&String> for CustomSymbol<N, A> {
 
 impl<const N: usize, A: Alphabet<N>> From<CustomSymbol<N, A>> for String {
     fn from(value: CustomSymbol<N, A>) -> Self {
-        let mut n = value.data;
-        let mut chars: Vec<char> = Vec::new();
-        let len = (A::ALPHABET.len() + 1) as u128;
-        loop {
-            let i = n % len;
-            n -= i;
-            n /= len;
-            chars.push(A::ALPHABET[i as usize - 1]);
-            if n == 0 {
-                break;
-            }
+        let mut rem = value.data;
+        let char_size = (A::ALPHABET.len() + 1) as u128;
+        let mut result = String::with_capacity(A::MAX_SYMBOL_LEN);
+        while rem != 0 {
+            let it = rem % char_size;
+            rem -= it;
+            rem /= char_size;
+            result.push(A::ALPHABET[it as usize - 1]);
         }
-        chars.into_iter().rev().collect()
```

Comment:
This is marginally better because it removes `log(len)` re-allocations of `Vec` bc it's zero-allocated at the beginning.

I used String as it's a bit clearer and turned `loop` into `while` with condition, as well as renaming variables.

I was actually trying to remove `.into_iter().rev().collect()` but there's no way of shrinking the `String` from front without a bunch of manual memory management and it ends up causing fragmentation so I guess there's no better way.

What about reversing the order of `u128`? Reverse encoding costs the same but decoding doesn't have to do this move.

Reply:
reversing the u128 could be fruitful

Reply:
I did that, and reversing input string iteration _should_ be faster as the input &str characters are random access due to std encoding being utf-8 so `.chars().rev()` should be optimized into a reverse bytes iterator.

This means `result` can be built directly without reversing the memory which means there's only a single allocation left now, and in case of Display, a copy to target buffer.